src/libraries/System.Net.Security/src/System/Net/StreamFramer.cs

```diff
@@ -31,24 +31,17 @@ internal sealed class StreamFramer
 
             byte[] buffer = _readHeaderBuffer;
 
-            int bytesRead;
-            int offset = 0;
-            while (offset < buffer.Length)
+            int bytesRead = await TAdapter.ReadAtLeastAsync(
```

Comment:
There is a risk here.  Previously if the async read pends, it would force the ReadMessageAsync state machine to be allocated.  Now if the async read pends, it will force the ReadMessageAsync state machine to be allocated _and_ the ReadAtLeastAsync state machine to be allocated.

We should consider using the pooled async state machine builders on the ReadAtLeastAsync async methods.

Reply:
Is this as simple as adding

```C#
[AsyncMethodBuilder(typeof(PoolingAsyncValueTaskMethodBuilder))]
```

to the `Stream.ReadAtLeastAsyncCore` method?  Or is it more involved than that?

Reply:
Implementation-wise, that's it, easy peasy.  The important thing is validating performance.  The concern with such pooling is that it can have a negative impact on the larger system, especially if it ends up creating more gen2 to gen0 references (by referencing gen0 objects from pooled objects that are likely to eventually be gen2).  But I worry that without doing this, the methods will end up not being used because folks will see the allocation and choose to avoid the helpers as a result.

Reply:
Let's chat tomorrow about how I can validate the performance here. I'd like to get your thoughts on what to try / look for.